ExternalProject SUNDIALS 3.2.1

alexsavulescu opened the pull request:
* include work from cvode3 branch

closes #1328

- [x]  Cvode version 3 test data must be visually compared to the legacy data for acceptable similarity.
- [ ]  Must use nrn-modeldb-ci to visually compare the differences.
- [x]  test coverage substantially complete
- [x]  merge rxd testdata and update submodule commit

alexsavulescu committed:
ExternalProject SUNDIALS 3.2.1
* include work from cvode3 branch

alexsavulescu committed:
ninja needs BUILD_BYPRODUCTS in ExternalProject_Add

alkino committed:
Merge branch 'master' into slds

alkino committed:
Format

alkino committed:
Merge remote-tracking branch 'origin/master'

alkino committed:
Format

alkino committed:
Fix name of SUNR functions

nrnhines committed:
Fix segfault (missing IDASetUserData) and N_VSpace_... prototypes.

alkino committed:
Build sundials with the same type than nrn

alkino committed:
Fully remove sundials

alkino committed:
Merge remote-tracking branch 'origin/master'

nrnhines committed:
IDASetStopTime should not be called for interpolation.

nrnhines committed:
ctest -R netpar passes. MPI_Comm not available from cvodeobj.cpp

nrnhines commented:
@alkino Just to let you know, I have a fix for ```ctest -R netpar``` but need to extend to NRNMPI_DYNAMICLOAD.

alkino commented:
> @alkino Just to let you know, I have a fix for `ctest -R netpar` but need to extend to NRNMPI_DYNAMICLOAD.

NRNMPI_DYNAMICLOAD should be cleaned to be the same than in coreneuron (double implentation today): https://github.com/neuronsimulator/nrn/issues/2285, so feel free to rework everything

nrnhines committed:
Revert "ctest -R netpar passes. MPI_Comm not available from cvodeobj.cpp"

This reverts commit e0b3ef1928a04c18b455455d1d8c49e96676b8c0.

nrnhines committed:
ctest -R netpar passes. Works with NRNMPI_DYNAMICLOAD.

nrnhines committed:
Problem translating long long foo for NRNMPI_DYNAMICLOAD

nrnhines committed:
update nrntest (cvode version 3 and no hhtable)

alkino committed:
Merge branch 'master' into slds

nrnhines committed:
Selected messages from old digest-debug branch

Print digest of cvode f(y,t) and solvex(b,...)
  cmake -DNRN_DIGEST=ON
  nrn_digest() starts the accumulation of cvode digest info.
  nrn_digest("filename") prints accumulated cvode digest info.
  filename format is:
    message threadid t t(in exact hex format) digest
    where digest is the first 16 hex characters of the SHA1 hash
    of the double* array indicated by the message.

-DNRN_ENABLE_ARCH_INDEP_EXP_POW=ON (default OFF)
  Provides h.use_exp_pow_precision(style)
  style = 0 means use normal machine IEEE precision for exp(x) and pow(x,y)
  style = 1 means use 53 bit mpfr
  style = 2 means use IEEE but truncate to 32 bit precision.

   sundials uses hoc_pow.

   digest format has a f(y,t) call count.

Use 32bit mantissa round for hoc_Exp (as well as hoc_pow).

See if pow is the culprit (round mantissa to 32 bit accuracy).

print y and ydot for first t>0

%a format different between linux and windows

nrnhines committed:
libcrypt for mac

nrnhines committed:
Merge branch 'master' into slds

nrnhines committed:
No need link libsundials_nvecparallel.a

nrnhines committed:
Merge branch 'master' into slds

nrnhines committed:
format

nrnhines committed:
Left out }

nrnhines committed:
Revert "No need link libsundials_nvecparallel.a"

This reverts commit 824732b48c4dbb4794426ee99191dba45bebdd9f.

nrnhines committed:
Merge branch 'master' into slds

nrnhines committed:
"major.minor.patch" = h.CVode().version()

nrnhines committed:
eliminate pytest warning

nrnhines committed:
CVode 3.2.1 json data for test_nrntest_fast.py

alkino committed:
Merge branch 'master' into slds

alkino committed:
Merge remote-tracking branch 'origin/master'

alkino committed:
sundials nvecparallel needs MPI

alkino committed:
Merge remote-tracking branch 'origin/slds'

alkino committed:
Fix FindSUNDIALS.cmake

alkino committed:
fix linkage

alkino committed:
Try to clean linkage with sundials

alkino committed:
Format

alkino committed:
Don't use corenrn if corenrn not set

alkino committed:
Fix fix and fix

nrnhines commented:
It is difficult getting tests to pass in a principled manner with respect to quantitative identity of results. Differences are sometimes seen with respect to architecture, operating system, floating point differences between transcendental functions (pow(x,y) for example), compiler version, order of summation (eg. threads, MPI), and, now, cvode version.  I'm guessing that it might be worthwhile to  use a test strategy that does a series of cvode runs with a series of a absolute tolerances at fixed exact time points and event times and verify that the  "results" approach an asymptotic standard result.

I've gotten cvodeobj.cpp to compile with Sundials v 6 and when the whole thing compiles will see if I can get some tests such as test_nrntest_fast.py modified to pass for cv2, 3, 6 on windows/mac/linux

alkino committed:
Format

alkino commented:
> It is difficult getting tests to pass in a principled manner with respect to quantitative identity of results. Differences are sometimes seen with respect to architecture, operating system, floating point differences between transcendental functions (pow(x,y) for example), compiler version, order of summation (eg. threads, MPI), and, now, cvode version. I'm guessing that it might be worthwhile to use a test strategy that does a series of cvode runs with a series of a absolute tolerances at fixed exact time points and event times and verify that the "results" approach an asymptotic standard result.
> 
> I've gotten cvodeobj.cpp to compile with Sundials v 6 and when the whole thing compiles will see if I can get some tests such as test_nrntest_fast.py modified to pass for cv2, 3, 6 on windows/mac/linux

Should we pass to sundials 6 directly?

alkino committed:
trick depends

nrnhines commented:
> Should we pass to sundials 6 directly?

I don't think so. Undoubtedly there are bugs in my modifications.  At this point, it is just another example see if we can eliminate
version specific data for CI tests.

alkino committed:
try to help sundials to find MPI

alkino commented:
`/root/nrn/external/sundials/lib64/libsundials_ida.a`

`lib` vs `lib64` from sundials, I don't know how to know that from findSUNDIALS.cmake

alkino committed:
set LIBDIR

pramodk commented on setup.py:
```diff
@@ -419,9 +419,15 @@ def setup_package():
-                "-DNMODL_ENABLE_PYTHON_BINDINGS=ON",
             ]
+            + (
+                [
+                    "-DCORENRN_ENABLE_OPENMP=ON",  # TODO: manylinux portability questions
+                    "-DNMODL_ENABLE_PYTHON_BINDINGS=ON",
+                ]
+                if Components.CORENRN
+                else []
```
this is not relevant here? (or, rebase is missing?)

alkino replied:
OK, I splitted this small fix: https://github.com/neuronsimulator/nrn/pull/2368

alkino committed:
-tp=haswell is not accepted by sundials

alkino committed:
Merge branch 'master' into slds

alkino committed:
Merge remote-tracking branch 'origin/master'

alkino committed:
Remove too more spaces

alkino committed:
Merge remote-tracking branch 'origin/master'

alkino committed:
CVodeSetFdata => CVodeSetUserData

nrnhines committed:
Merge branch 'master' into slds

nrnhines committed:
eliminate some compiler warnings

nrnhines committed:
update nrntest to master merged hines/update-t12

nrnhines committed:
Merge branch 'master' into slds

nrnhines committed:
on previous master merge, resolve conflict incorrectly

nrnhines committed:
Merge branch 'master' into slds

alkino committed:
Merge remote-tracking branch 'origin/master'

nrnhines committed:
Merge branch 'master' into slds

nrnhines committed:
some files missing in NRN_PARALLEL_SRC_FILES

nrnhines committed:
ninja format

nrnhines committed:
Merge branch 'master' into slds (master for test_nlayer.py conflict)

nrnhines committed:
cvode.version() got lost in a merge

nrnhines committed:
test_kschan data for cvode 3.2.1

nrnhines committed:
test_nrntest_fast.py tolerance adjustment for t14

nrnhines committed:
missing json file for ctest -R parallel::nrntest_fast

nrnhines committed:
cv3 version of json file for test_netcvode.py

nrnhines committed:
Merge branch 'master' into slds

nrnhines committed:
temporarily turn off CI tests of rxd

nrnhines committed:
temporarily turn off (most) wheel builds and gitlab CI

nrnhines committed:
adust a few tolerance values to see if more CI pass

nrnhines committed:
Merge branch 'digest-debug-2' into digest-debug-3

nrnhines committed:
nrn_digest(i) will print the details of the i'th digest item.

nrnhines committed:
missing include

nrnhines committed:
Allow digest when multiple threads.

nrnhines committed:
digest print solvex enter y

nrnhines committed:
sundial RPowerR uses hoc_pow. nrn_digest(i, "abort") calls abort() on
reaching index i

nrnhines committed:
Documentation for nrn_digest and use_exp_pow_precision

nrnhines committed:
nrn_digest signature consistent with documentation

nrnhines committed:
occvode.cpp -Wall free

nrnhines committed:
Merge branch 'master' into hines/digest-debug-3

nrnhines committed:
The configure option is NRN_ENABLE_ARCH_INDEP_EXP_POW

nrnhines committed:
format

nrnhines committed:
Merge branch 'master' into hines/digest-debug-3

nrnhines committed:
Revert "adust a few tolerance values to see if more CI pass"

This reverts commit 21c7826a130f7ae951a887d70f5d8feec98a7a4a.

nrnhines committed:
Revert "temporarily turn off (most) wheel builds and gitlab CI"

This reverts commit 0d4cc21658003e93535347cabb3f1cef928ec22e.

nrnhines committed:
Revert "temporarily turn off CI tests of rxd"

This reverts commit eb5e85a22bb4f56e235317aa787021962a88db89.

nrnhines committed:
Merge branch 'master' into slds

nrnhines committed:
Merge branch 'hines/digest-debug-3' into hines/cv3-digest

nrnhines committed:
Use the neuronsimulator fork of sundials

nrnhines committed:
passes ctest -j 8 (with "-ffp-model=strict")

nrnhines committed:
Merge branch 'master' into hines/cv3-digest

nrnhines committed:
Merge branch 'master' into hines/cv3-digest

nrnhines committed:
Update test/rxd/testdata for cvode3

nrnhines committed:
document how to update rxd test data

nrnhines committed:
Revert "Update test/rxd/testdata for cvode3"

This reverts commit 7c670b7b27bb85091367ea4c42d6caea74820894.

nrnhines committed:
fix typo

nrnhines committed:
update rxdtest repository (e0a6bc281) 2024-10-22

nrnhines committed:
Merge branch 'master' into hines/cv3

nrnhines committed:
cd share/lib/python/neuron/tests; python test_rxd.py --save test_rxd.json

nrnhines committed:
extend the rxd test data instructions

nrnhines committed:
propagate compiler from nrn to sundials. SUNDIALS MPI_ENABLE=OFF

nrnhines committed:
Merge branch 'master' into hines/cv3

nrnhines committed:
Since sundials MPI_ENABLE=OFF have our own nvector_parallel.h

nrnhines committed:
nvector_parallel.cpp agnostic with respect to MPI_Comm

nrnhines committed:
cast between sunindextype and long

nrnhines committed:
update a cv3.json file.

nrnhines committed:
nrnparallel_ld follows nvector_parallel style for MPI_Comm

nrnhines committed:
relax resolution requirements so macosx will pass CI without needing
-DNRN_ENABLE_ARCH_INDEP_EXP_POW=ON
-DCMAKE_C_FLAGS="-ffp-model=strict" -DCMAKE_CXX_FLAGS="-ffp-model=strict"

nrnhines committed:
adjust tol of test_nrntest_fast.py::test_t14[cvode_long_double-3-t]
  needed for 4 ubuntu CI. Not sure about the underlying cause.

nrnhines committed:
black format for test/.../*.py

pramodk commented on cmake/modules/FindSUNDIALS.cmake:
```diff
@@ -0,0 +1,58 @@
+  # cmake-format: off
+  # GIT_REPOSITORY https://github.com/LLNL/sundials.git
+  # Forked to following, to allow RPowerR to use hoc_pow
+  GIT_REPOSITORY https://github.com/neuronsimulator/sundials.git
+  # GIT_TAG e2f29c34f324829302037a1492db480be8828084   6.2.0 -> CVodeMem no longer "visible" GIT_TAG
+  # c09e732080a214694b209032ec627c93fed45340  4
+  #GIT_TAG 811234254d37652954daff0ccdb7af9813736846 # 3.2.1
+  GIT_TAG 71b3daec18ce66a7c011be9501a31567aa5c09b6 # 3.2.1 hines/hoc_pow
+  # GIT_TAG 7ed895fd102226dbc52225e6b2c6e26ef1cafa0e  #2.7.0
+  # cmake-format: on
```
@nrnhines : is this https://github.com/LLNL/sundials/compare/main...neuronsimulator:sundials:hines/hoc_pow only change that we need to have? Indeed it's look very trivial.

nrnhines replied:
That is correct for Sundials version 3.  We'll eventually see if changes need to be more substantial for later versions. There is some question about where to put the 10 nrn/src/nrniv/nvector_* files. They distinguish between (serial, thread, mpi) and (double, longdouble) and I don't think the changes are relevant to sundials as they have details that depend on how NEURON uses threads and mpi and NRN_ENABLE_MPI_DYNAMIC. At the moment, when
NRN_ENABLE_MPI_DYNAMIC=ON, then ```nm libnrniv.so | grep MPI``` shows nothing that points into libmpi. Instead, many things point into mpi specific interface libraries such as libnrnmpi_mpich.so and libnrnmpi_ompi.so

pramodk commented on cmake/modules/FindSUNDIALS.cmake:
```diff
@@ -0,0 +1,58 @@
+  BUILD_BYPRODUCTS
+    <INSTALL_DIR>/lib/libsundials_ida.a <INSTALL_DIR>/lib/libsundials_cvode.a
+    <INSTALL_DIR>/lib/libsundials_nvecserial.a <INSTALL_DIR>/lib/libsundials_nvecpthreads.a
+    <INSTALL_DIR>/lib/libsundials_nvecparallel.a)
```
This was previously discussed with Alex, but still remember less-than-ideal experiences with `ExternalProject_Add` during 2016-17. I am still wondering if there might be better alternatives to using nested cmake configure steps. :) 

I'll leave it to @alkino and @1uc to weigh in.

nrnhines replied:
My experience with ```ExternalProject_Add``` for Sundials hasn't been too bad. At least any change to a build at the nrn level propagates to re-compile and build the sundials *.a files (that link into libnrniv.so).  It could be better in terms of having a build tree as a subtree in the nrn build tree. (I often use mulitple nrn/build.../ with different cmake option choices and it would be nice if the latest nrn build did not cause the previous sundial build to be overwritten.  I would also prefer to avoid ```git status``` not showing the untracked folder ```external/sundials``` At the moment the sundials repository seems to begin at nrn/external/sundials/src/sundials-external, the build takes place in nrn/external/sundials/src/sundials-external-build, and the install takes place in nrn/external/sundials/(include, lib).  I suppose the way we have been handling InterViews and, more recently, fmt, might provide an alternative strategy.

alexsavulescu replied:
I'm all in for ExternalProject_Add 🚀 given that we can do custom build/configure commands, it will be definitely useful for upgrading to newer sundials.

pramodk commented on src/nrniv/nvector_nrnparallel_ld.cpp:
```diff
@@ -15,19 +15,20 @@
 #include "nvector_nrnparallel_ld.h"
-#include "sundialsmath.h"
-#include "sundialstypes.h"
+#define MPI_Comm int  // or perhaps get rid of them all
```
nitpick - just curious about this

nrnhines replied:
I am seriously considering removing all references to MPI_Comm in nrn/src/nrniv nvector_parallel.* and nvector_nrnparallel_ld.*  . These are basically copies of sundials files from version 2 and 3 but modified to allow
construction from within nrn/src/nrncvode/cvodeobj.cpp and also work without directly calling anything from within mpi.h. That is, they are valid for ```-DNRN_ENABLE_MPI_DYNAMIC= ON or OFF and don't involve any mpi.h structures (such as MPI_Comm) or functions. They only call things available from ```nrn/src/nrnmpi/nrnmpi.h```.  I left in the MPI_Comm tokens just as indicators of what might have to be changed in future movment toward Sundials version 7.  At the moment, all of SUNDIALS used by neuron is inside libnrniv.so

pramodk reviewed (commented):
Glad to see this is close to becoming a reality!

I didn’t have time to review the C++ changes, but I think @alkino can help with that.

I have just added two minor comments regarding the build parts.

nrnhines commented:
@pramodk I was looking at the checkbox you added to the top commit
merge rxd testdata and update submodule commit
and was wondering what is meant.  The nrn master already has
```
commit f8e94c53da9d59570e6922acef6715f85776ff2e
Author: adamjhn <adam.newton@yale.edu>
Date:   Thu Oct 24 14:59:12 2024 -0400

    Move reference data from test_rxd.py to test_rxd.json. (#3143)
    
    
    * Added a --save flag to regenerate the data.
 ```
and the sundials version that this PR uses is mentioned in nrn/cmake/modules/FindSUNDIALS.cmake
```
GIT_REPOSITORY https://github.com/neuronsimulator/sundials.git
GIT_TAG 71b3daec18ce66a7c011be9501a31567aa5c09b6 # 3.2.1 hines/hoc_pow
```
So it seems the checkbox should have an [x]

alexsavulescu commented on cmake/modules/FindSUNDIALS.cmake:
```diff
@@ -0,0 +1,58 @@
+             -DBUILD_SHARED_LIBS=OFF
+             -DMPI_C_COMPILER=${MPI_C_COMPILER}
+             -DCMAKE_INSTALL_LIBDIR=${SUNDIALS_LIB_DIR}
+  BUILD_BYPRODUCTS
```
FYI: this is needed for Ninja

alexsavulescu commented on src/nrncvode/cvodeobj.cpp:
```diff
@@ -1442,36 +1477,49 @@ void Cvode::statistics() {
+        assert(flag == CVDLS_SUCCESS);
+
+        /* Set jacobian and allocate memory */
+        /* (TODO Michael Hines: is there a jacobian for dense matrix
```
I did not review this at the time of the initial PR

nrnhines replied:
We are using the default built-in jacobian from sundials. I changed the comment to mention that.

nrnhines committed:
Changes after ctest coverage analysis.

nrnhines committed:
Merge branch 'master' into slds

nrnhines committed:
Replace cvode.extra_scatter_gather with h.beforestep_callback

nrnhines committed:
Replace cvode.extra_scatter_gather with h.beforestep_callback

nrnhines committed:
rxd tests use rxdtestdata hines/data-fix af269297

nrnhines committed:
rxd/testdata when test/rxd/3d/test_multi_gridding_mix.py not skipped

nrnhines committed:
Merge branch 'master' into hines/cv3

nrnhines committed:
use rxdtestdata hines/cvode3 721ac55a

nrnhines committed:
Don't use cvode version specific filenames or keys for json data.

nrnhines committed:
ecs_diffusion needs section to gather data.

nrnhines committed:
Merge branch 'master' into hines/rxd-data-fix

nrnhines committed:
Merge branch 'hines/rxd-data-fix' into hines/cv3 and update testdata

nrnhines committed:
Merge branch 'master' into hines/cv3

nrnhines committed:
Merge branch 'master' into slds

 Conflicts:
	docs/python/simctrl/cvode.rst
	test/cover/test_netcvode.py
	test/external/CMakeLists.txt
	test/pytest_coreneuron/test_nrntest_fast.py
	test/rxd/testdata

nrnhines committed:
cmake: update nrniv_lib linking for SUNDIALS v3 submodule

- Use keyword signatures consistently in target_link_libraries
- Make SUNDIALS include directory PRIVATE + BUILD_INTERFACE only

nrnhines committed:
build: remove obsolete SUNDIALS sources

The SUNDIALS library is now provided via CMake submodule (external/sundials).
The src/sundials directory is no longer needed.

nrnhines committed:
format

nrnhines committed:
test_neurondemo comparison data for cover3454-cv3

nrnhines committed:
build: add SUNDIALS as git submodule (neuronsimulator fork)

- Add external/sundials submodule pointing to neuronsimulator/sundials (hines/hoc_pow)
- Update FindSUNDIALS.cmake to use the standard submodule approach.
- Build SUNDIALS inside the main build directory.
- Initialization consistent with other submodules (InterViews, etc.)

nrnhines committed:
format

nrnhines committed:
update coding-conventions (prints black version)

nrnhines committed:
style: reformat rxd files with black under Python 3.10

Matches formatting used in CI (Python 3.10.12 + black 22.12.0).

nrnhines committed:
build: make format-pr more robust

- Support both 'main' and 'master' as base branch
- Only pass existing files to the formatter
- Eliminates noisy "ERROR: X jobs failed" when files were deleted

nrnhines committed:
Compare two RxD testdata directories pairwise
* see test/rxd/compare_rxd_testdata_dirs.py -h

nrnhines committed:
update rxdtestdata to hines/slds-data b8d97dd92b

nrnhines committed:
Merge branch 'master' into slds-wt
